pkg/gateway/gateway.go

```diff
@@ -0,0 +1,206 @@
+	return func(gw Gateway, o *gatewayOptions) error {
+		gw.(*gateway).sdk = sdk
+
+		cfg, err := sdk.Config()
+
+		if err != nil {
+			return errors.Wrap(err, "Unable to access SDK configuration")
+		}
+
+		value, ok := cfg.Lookup("client.organization")
+		if !ok {
+			return errors.New("No client organization defined in the config")
+		}
+		gw.(*gateway).org = value.(string)
+
+		return nil
+	}
+}
+
+// WithIdentity is an optional argument to the Connect method which specifies
+// the identity that is to be used to connect to the network.
+// All operations under this gateway connection will be performed using this identity.
+func WithIdentity(wallet Wallet, label string) IdentityOption {
+	return func(gw Gateway, o *gatewayOptions) error {
+		mspClient, err := msp.New(gw.getSdk().Context(), msp.WithOrg(gw.getOrg()))
+		if err != nil {
+			return err
+		}
+
+		creds, err := wallet.Get(label)
+		if err != nil {
+			return err
+		}
+
+		var identity mspProvider.SigningIdentity
+		switch v := creds.(type) {
+		case *X509Identity:
+			identity, err = mspClient.CreateSigningIdentity(mspProvider.WithCert([]byte(v.GetCert())), mspProvider.WithPrivateKey([]byte(v.GetKey())))
+			if err != nil {
+				return err
+			}
+		}
+
+		o.Identity = identity
+		return nil
+	}
+}
+
+// WithUser is an optional argument to the Connect method which specifies
+// the identity that is to be used to connect to the network.
+// All operations under this gateway connection will be performed using this identity.
+func WithUser(user string) IdentityOption {
+	return func(gw Gateway, o *gatewayOptions) error {
+		o.User = user
+		return nil
+	}
+}
+
+// WithCommitHandler is an optional argument to the Connect method which
+// allows an alternative commit handler to be specified. The commit handler defines how
+// client code should wait to receive commit events from peers following submit of a transaction.
+// Currently unimplemented.
+func WithCommitHandler(handler CommitHandlerFactory) Option {
+	return func(gw Gateway, o *gatewayOptions) error {
+		o.CommitHandler = handler
+		return nil
+	}
+}
+
+// WithDiscovery is an optional argument to the Connect method which
+// enables or disables service discovery for all transaction submissions for this gateway.
+func WithDiscovery(discovery bool) Option {
+	return func(gw Gateway, o *gatewayOptions) error {
+		o.Discovery = discovery
+		return nil
+	}
+}
+
+func (gw *gateway) getSdk() *fabsdk.FabricSDK {
+	return gw.sdk
+}
+
+func (gw *gateway) getOrg() string {
+	return gw.org
+}
+
+func (gw *gateway) getPeersForOrg(org string) ([]string, error) {
+	value, ok := gw.cfg.Lookup("organizations." + org + ".peers")
+	if !ok {
+		return nil, errors.New("No client organization defined in the config")
+	}
+
+	val := value.([]interface{})
+	s := make([]string, len(val))
+	for i, v := range val {
+		s[i] = fmt.Sprint(v)
+	}
+
+	return s, nil
+}
+
+func (gw *gateway) GetNetwork(name string) (Network, error) {
```

Comment:
IMO, It would be more idiomatic Go to return a struct pointer rather than an interface in this case.